c8d: Add `platform` parameter to history, save and load

vvoland opened the pull request:
**- What I did**
Add the ability to select a specific platform for `docker history`, `docker save` and `docker load`.

**- How I did it**
See commits.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
`GET /images/{name}/history` now supports a `platform` parameter (JSON encoded OCI Platform type) that allows to specify a platform to show the history of.
`POST /images/{name}/load` and `GET /images/{name}/get` now support a `platform` parameter (JSON encoded OCI Platform type) that allows to specify a platform to load/save. Not passing this parameter will result in loading/saving the full multi-platform image.
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland commented on client/interface.go:
```diff
@@ -91,18 +91,25 @@ type ImageAPIClient interface {
-	ImageSave(ctx context.Context, images []string) (io.ReadCloser, error)
 	ImageTag(ctx context.Context, image, ref string) error
 	ImagesPrune(ctx context.Context, pruneFilter filters.Args) (image.PruneReport, error)
+
+	// TODO: Change variadic opts into a single parameter
+	ImageHistory(ctx context.Context, image string, opts ...ImageHistoryOptions) ([]image.HistoryResponseItem, error)
+
+	// TODO: Change variadic opts into a single parameter
+	ImageSave(ctx context.Context, imageIDs []string, opts ...ImageSaveOptions) (io.ReadCloser, error)
+
+	// TODO: Replace ImageLoad with ImageLoadVNext
+	ImageLoad(ctx context.Context, input io.Reader, quiet bool) (image.LoadResponse, error)
+	ImageLoadVNext(ctx context.Context, input io.Reader, opts ImageLoadOptions) (image.LoadResponse, error)
```
Any better way to do it? 🙈 

cc @thaJeztah @cpuguy83

thaJeztah replied:
Should we put these in the same location as the other options? (`image.LoadOptions`), or won't that work?

vvoland replied:
That would work, was considering it, but I think that (all) of them should really be separate types.

Sharing the same options structs between the client and the backend implementation in the daemon tends to leak the implementation details from the image service into the shared types.

vvoland replied:
On the other hand.. I think it's actually the backend that does the bad thing here by reusing these. And Options struct in the `api/types/image` are actually expected to be consumed by the go client, and not the daemon's backend.

vvoland replied:
Moved to `api/types/image`

vvoland replied:
Ah, this will still break the interface matching for consumers 😭

vvoland commented:
Unrelated failure, looks flaky, but haven't seen that one yet.

```
=== FAIL: github.com/docker/docker/container TestStateTimeoutWait (0.22s)
    state_test.go:147: Stop callback doesn't fire in 200 milliseconds
```

vvoland committed:
image/history: Support `Platform` parameter

Add `Platform` parameter that allows to select a specific platform to
show the history for.

This is a breaking change to the Go client as it changes the signature
of `ImageHistory`.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
image/save&load: Support `Platform` parameter

Add `Platform` parameter that allows to select a specific platform to
save/load.

This is a breaking change to the Go client as it changes the signatures
of `ImageLoad` and `ImageSave`.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on client/image_load_test.go:
```diff
@@ -9,6 +9,8 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/docker/docker/api/types/image"
+
 	"github.com/docker/docker/errdefs"
```
Stray empty line here

thaJeztah commented on client/image_load.go:
```diff
@@ -12,12 +13,28 @@ import (
+
+		p, err := json.Marshal(*opts.Platform)
+		if err != nil {
+			return image.LoadResponse{}, err
```
Should we return a `errdefs` error here? Also probably would be good to wrap the error as we did for the other one (`invalid platform:`)

vvoland replied:
Not sure - other usages of `json`'s `Marshal/Unmarshal/NewDecoder/NewDecoder` don't return a raw error directly so I went with the same approach.

cpuguy83 replied:
Seems fine not to wrap it.

thaJeztah replied:
Yeah, we could probably make a pass at typing those errors.

As to the wrapping; it was because I noticed that for ImageHistory we add a prefix;


```go
return nil, fmt.Errorf("invalid platform: %v", err)
```

Without that, the error may not give a clue where the problem is (just a plain JSON marshal error).

thaJeztah commented on client/image_save.go:
```diff
@@ -2,17 +2,33 @@ package client // import "github.com/docker/docker/client"
+
+		p, err := json.Marshal(*opts.Platform)
+		if err != nil {
+			return nil, fmt.Errorf("invalid platform: %v", err)
```
Same here for considering to use an `errdefs`.


☝️ Actually starting to wonder if it would make sense to have a utility for this; yes, it would be minimal (just json marshalling and handling errors), but would help producing the right errors and being consistent WDYT?


☝️ And I just realise we have `httputils.DecodePlatform`, so to some extent it would be nice if we had the reverse to that, but I guess `httputils` is part of `api/server`, so if we do, we should look where a suitable location would be if we want to put them together 🤔

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -85,7 +83,17 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
+		if platform != nil {
+			newTarget, err := i.getPushDescriptor(ctx, img, platform)
+			if err != nil {
+				return errors.Wrap(err, "no suitable export target found")
```
- Would it make sense to include `platform` here somehow (to make clear that there's no image found for the given platform)? Or does `getPushdescriptor` already return this?
- Do we need an `errdefs`  type here?

vvoland replied:
Ah, good catch. Added platform in message here.
For `errdefs` - `getPushDescriptor` is expected to return an errdefs already, fixed a case where it didn't.

thaJeztah commented on daemon/images/image_exporter.go:
```diff
@@ -4,17 +4,24 @@ import (
 	"context"
 	"io"
 
+	"github.com/docker/docker/errdefs"
+	"github.com/pkg/errors"
+
 	"github.com/containerd/log"
```
Stray empty line

thaJeztah commented on daemon/images/image_exporter.go:
```diff
@@ -4,17 +4,24 @@ import (
-func (i *ImageService) ExportImage(ctx context.Context, names []string, outStream io.Writer) error {
+func (i *ImageService) ExportImage(ctx context.Context, names []string, platform *ocispec.Platform, outStream io.Writer) error {
+	if platform != nil {
+		return errdefs.NotImplemented(errors.New("platform parameter is not supported with the graphdriver image store"))
```
This probably should be a `ErrInvalidParameter` here, otherwise the API would return a `501`.


Well, either that, or (maybe more "correct"?) we must somehow catch uses of `platform` in the API if the containerd image-store is not enabled, and return an "invalid parameter" there 🤔

vvoland replied:
Not needed after implementing it! 😅

thaJeztah commented on daemon/images/image_exporter.go:
```diff
@@ -44,7 +51,10 @@ func (i *ImageService) PerformWithBaseFS(ctx context.Context, c *container.Conta
 // ball containing images and metadata.
-func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error {
+func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platform *ocispec.Platform, outStream io.Writer, quiet bool) error {
+	if platform != nil {
+		return errdefs.NotImplemented(errors.New("platform parameter is not supported with the graphdriver image store"))
+	}
```
Same here


☝️ ❓ wondering now - could we somehow actually implement this? i.e. _if_ a platform is provided and it matches then proceed - if it doesn't matches, then produce the same error as we do if we have multi-arch support (containerd image store) enabled?

thaJeztah commented on integration/build/build_squash_test.go:
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/docker/docker/api/types/image"
+
 	"github.com/docker/docker/api/types"
```
Stray empty line

thaJeztah commented on integration/image/save_test.go:
```diff
@@ -13,6 +13,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/docker/docker/api/types/image"
+
 	"github.com/cpuguy83/tar2go"
```
Stray empty line

thaJeztah commented on internal/testutils/specialimage/load.go:
```diff
@@ -9,6 +9,8 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/docker/docker/api/types/image"
+
 	"github.com/docker/docker/client"
```
Stray empty line

thaJeztah commented on api/server/router/image/backend.go:
```diff
@@ -32,9 +32,9 @@ type imageBackend interface {
 type importExportBackend interface {
-	LoadImage(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error
+	LoadImage(ctx context.Context, inTar io.ReadCloser, platform *ocispec.Platform, outStream io.Writer, quiet bool) error
 	ImportImage(ctx context.Context, ref reference.Named, platform *ocispec.Platform, msg string, layerReader io.Reader, changes []string) (dockerimage.ID, error)
-	ExportImage(ctx context.Context, names []string, outStream io.Writer) error
+	ExportImage(ctx context.Context, names []string, platform *ocispec.Platform, outStream io.Writer) error
```
Starting to wonder if we should (also) have a options-type defined for these endpoints, as the list of arguments is starting to grow.

vvoland replied:
I think it's still fine, as it's only one parameter that is a truly "option" (inTar and outStream aren't really "options").
It's also only an internal interface, so we don't need to worry about breaking compatibility in future.
I don't think we're gonna add another parameter any soon, and if we do, then we can extract it to a struct.

thaJeztah reviewed (changes requested):
Did a pass, but happy to hear your thoughts on some of them.

laurazard commented on api/swagger.yaml:
```diff
@@ -9936,7 +9945,16 @@ paths:
+            JSON encoded OCI platform describing a platform which will be used
+            to select a platform-specific image to be saved if the image is
+            multi-platform.
+            If not provided, full multi-platform image will be saved.
```
```suggestion
            If not provided, the full multi-platform image will be saved.
```

vvoland replied:
Thanks!

laurazard commented on api/swagger.yaml:
```diff
@@ -10003,6 +10021,16 @@ paths:
+            JSON encoded OCI platform describing a platform which will be used
+            to select a platform-specific image to be load if the image is
+            multi-platform.
+            If not provided, full multi-platform image will be loaded.
```
```suggestion
            If not provided, the full multi-platform image will be loaded.
```

laurazard reviewed (approved):
LGTM (left a couple nits, and +1 on @thaJeztah's comments)

vvoland committed:
images: Support platform selection on graphdrivers

Allows load to filter image manifests to load based on their platform.
For save, verify that the image platform matches the requested platform,
otherwise error out.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

cpuguy83 commented on image/tarexport/tarexport.go:
```diff
@@ -24,10 +26,12 @@ type manifestItem struct {
+	lss             layer.Store
+	rs              refstore.Store
+	loggerImgEvent  LogImageEvent
+	platform        *platforms.Platform
```
Not sure platform is needed here since we have the matcher.

vvoland replied:
It's used in error message: https://github.com/moby/moby/blob/c25d20beb91db1a4811c0529eb8e5eba7dff5c7b/image/tarexport/save.go#L171

thaJeztah replied:
Yeah, would be great if matchers would expose more information about what they're matching and/or perhaps return an error instead of only have a boolean.

thaJeztah commented on api/common.go:
```diff
@@ -3,7 +3,7 @@ package api // import "github.com/docker/docker/api"
 const (
 	// DefaultVersion of the current REST API.
-	DefaultVersion = "1.47"
+	DefaultVersion = "1.48"
```
I opened a quick PR to update the version, as there's a couple of PRs pending that need a new version (and I was working on some); https://github.com/moby/moby/pull/48476

thaJeztah commented on api/swagger.yaml:
```diff
@@ -9202,6 +9202,15 @@ paths:
+            If not provided, the host platform will be used. If it's not
+            available, any present platform will be picked.
+
+            Example: {"os": "linux", "architecture": "arm", "variant": "v5"}
```
Non-blocking; wondering if it looks better with the example in back-ticks;

Without:

<img width="952" alt="Screenshot 2024-09-11 at 17 23 09" src="https://github.com/user-attachments/assets/99b901c2-dae7-49cb-8114-4b26ea09fb56">


With:

<img width="934" alt="Screenshot 2024-09-11 at 17 22 26" src="https://github.com/user-attachments/assets/538e85e1-6e68-44dc-8c2c-7c9fe98429f1">

vvoland replied:
Done!

thaJeztah commented on image/tarexport/save.go:
```diff
@@ -10,6 +10,8 @@ import (
 	"path/filepath"
 	"time"
 
+	"github.com/containerd/platforms"
+
 	"github.com/containerd/containerd/images"
```
🙈 stray empty line

vvoland replied:
Ugh, thanks! I really need to sort out my Goland settings 🙈

thaJeztah reviewed (approved):
LGTM

some minor comments for consideration
